### ISSUE TITLE
UseAzureAppConfiguration and AzureAppConfigurationRefreshMiddleware break context execution

### ISSUE THREAD

**Issue by fjbelizon** [Azure/AppConfiguration#915]
I have detected that if we make use of ‘UseAzureAppConfiguration’, which in turn internally injects the middleware ‘AzureAppConfigurationRefreshMiddleware’, it breaks the execution of the context of the Eternal Orchestration Functions.

I am using Azure Isolated Function in NET 8, using package ‘Microsoft.Azure.AppConfiguration.Functions.Worker’ in version 7.2.0.

The middleware internally performs an ‘ExecutionContext.SuppressFlow()’, which breaks the execution flow of the context: https://learn.microsoft.com/en-us/dotnet/api/system.threading.executioncontext.suppressflow?view=net-8.0

Right now the solution I am applying is to use ‘AddAzureAppConfiguration’ without ‘UseAzureAppConfiguration’. In my orchestrator constructor I inject ‘IConfigurationRefresherProvider’ to use the configuration refresher in the orchestrator activities.

The problem is not described anywhere and should be checked. Thanks in advance.

**Comment by amerjusupovic** [comment]
Hi @fjbelizon, we got a [similar issue](https://github.com/Azure/AppConfiguration-DotnetProvider/issues/484) about durable functions being broken a while back, but `ExecutionContext.SuppressFlow()` wasn't mentioned. In the discussion there as well as in the [linked issue](https://github.com/Azure/azure-functions-dotnet-worker/issues/1666) that's in the Azure Functions repo, it was thought that the problem was not specific to the Azure App Configuration middleware.

One helpful test would be to see if version 6.0.1 of the provider also causes this issue for you, since that was before we added the call to `SuppressFlow()`. If the issue goes away, I'll need to update that original issue. Thanks!

**Comment by fjbelizon** [comment]
Hi @amerjusupovic. Ok, I'm going to do the test

**Comment by fjbelizon** [comment]
@amerjusupovic yes, same problem. Version 6.0.1 also breaks the execution of Eternal Functions.

I attach the code sample for you to test: [bug-azure-app-configuration-refresh-middleware.zip](https://github.com/Azure/AppConfiguration/files/15319701/bug-azure-app-configuration-refresh-middleware.zip). I've put two ‘TODO’ tags for you to comment/uncomment the lines that make the execution work/fail.

I look forward to your review. Thanks in advance.

**Comment by sparraguerra** [comment]
Any news about this issue? Thanks in advance

**Comment by jimmyca15** [comment]
@amerjusupovic 

Given there was a repro with 6.0.1 do you believe it would be the same issue as [here](https://github.com/Azure/AppConfiguration-DotnetProvider/issues/484)?

**Comment by amerjusupovic** [comment]
@jimmyca15 Yes I believe so, I mentioned that someone on the functions team believed it wasn't an issue with the .NET provider library at first, but from this it seems like it is something we introduced. I'll look into this more and see if we can fix this portion of the code while still solving the [HttpContext issue](https://github.com/Azure/AppConfiguration-DotnetProvider/issues/424) that originally required this change.

**Comment by fjbelizon** [comment]
Ok, thanks @amerjusupovic 👌

**Comment by jimmyca15** [comment]
This looks like a dupe of [this issue](https://github.com/Azure/AppConfiguration-DotnetProvider/issues/484). I don't think `ExecutionContext.SuppressFlow` has any involvement here. This supported by 6.0.1 reproing as well.

**Comment by jimmyca15** [comment]
@amerjusupovic if you can confirm this is a dupe of the aforementioned issue we should merge this issue into the existing one.

**Comment by amerjusupovic** [comment]
Sorry for the confusion in my previous response, yes the repro in 6.0.1 suggests it's the same issue, and that was before the call to `SuppressFlow` was added. This should be merged with the existing one.

**Comment by BrunoCandia** [comment]
is there any news about this issue?, thanks in advance!!

**Comment by ivankaurov** [comment]
@BrunoCandia  we're using the `BackgroundService` that is thechnically doing the same stuff
```cs
internal sealed class AzureAppConfigRefreshService : BackgroundService
{
    private readonly IEnumerable<IConfigurationRefresher> _refreshers;

    public AzureAppConfigRefreshService(IConfigurationRefresherProvider refresherProvider)
    {
        _refreshers = refresherProvider.Refreshers;
    }

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        while (!stoppingToken.IsCancellationRequested)
        {
            try
            {
                foreach (var refresher in _refreshers)
                {
                    await refresher.TryRefreshAsync(stoppingToken);
                }

                await Task.Delay(...);
            }
            catch (OperationCanceledException) when (stoppingToken.IsCancellationRequested)
            {
                _logger.LogInformation("AzureAppConfig has been stopped");
                break;
            }
        }
    }
}
// ...
// serviceCollection.AddHostedService<AzureAppConfigRefreshService>();
```

**Comment by ivankaurov** [comment]
@amerjusupovic @jimmyca15 are there any news on this issue? We're still using our in-house kludge

**Comment by Fazer01** [comment]
We can totally relate to this issue! Following for updates :)

**Comment by kemmis** [comment]
Can we get an update on this? This was implemented which appears to be what your team required in order to fix this: https://github.com/Azure/azure-functions-dotnet-worker/issues/1666

**Comment by jimmyca15** [comment]
@kemmis I am following up on this. Ideally there should be no change required from the Azure App Configuration library.

**Comment by jimmyca15** [comment]
@kemmis Are you using the latest version of durable? From what I'm seeing, as long as you are using a late enough version this should work for you without any changes needed on our side. We can look to confirm this on our side by putting together a repro.

**Comment by kemmis** [comment]
@jimmyca15  Here are the packages I am using:
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net10.0</TargetFramework>
    <AzureFunctionsVersion>V4</AzureFunctionsVersion>
    <OutputType>Exe</OutputType>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

  <ItemGroup>
    <FrameworkReference Include="Microsoft.AspNetCore.App" />
    <PackageReference Include="Microsoft.Azure.Functions.Worker" Version="2.52.0" />
    <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.DurableTask" Version="1.18.0" />
    <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.EventGrid" Version="3.6.0" />
    <PackageReference Include="Microsoft.Azure.Functions.Worker.OpenTelemetry" Version="1.2.0" />
    <PackageReference Include="OpenTelemetry.Extensions.Hosting" Version="1.15.3" />
    <PackageReference Include="Azure.Monitor.OpenTelemetry.Exporter" Version="1.7.0" />
    <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.Http.AspNetCore" Version="2.1.0" />
    <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.OpenApi" Version="1.6.0" />
    <PackageReference Include="Microsoft.Azure.Functions.Worker.Sdk" Version="2.0.7" />
    <PackageReference Include="Microsoft.Azure.AppConfiguration.Functions.Worker" Version="8.5.0" />
  </ItemGroup>
```

And here is the exception I am getting:
```
System.InvalidOperationException: Orchestration history state was either missing from the input or not a string value.\r\n   at Microsoft.Azure.Functions.Worker.Extensions.DurableTask.Execution.DurableFunctionExecutor.RunOrchestrationAsync(FunctionContext context, BindingMetadata triggerBinding) in /_/src/Worker.Extensions.DurableTask/Execution/DurableFunctionExecutor.Orchestration.cs:line 29\r\n   at Microsoft.Azure.Functions.Worker.OutputBindings.OutputBindingsMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in /_/src/DotNetWorker.Core/OutputBindings/OutputBindingsMiddleware.cs:line 13\r\n   at Microsoft.Azure.AppConfiguration.Functions.Worker.AzureAppConfigurationRefreshMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next)\r\n   at Iwh.Ava.Functions.Middleware.InContactIdLoggingMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in E:\\agent1\\_work\\14\\s\\src\\Iwh.Ava.Functions\\Middleware\\InContactIdLoggingMiddleware.cs:line 37\r\n   at Iwh.Ava.Functions.Middleware.HttpRequestResponseLoggingMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in E:\\agent1\\_work\\14\\s\\src\\Iwh.Ava.Functions\\Middleware\\HttpRequestResponseLoggingMiddleware.cs:line 30\r\n   at Microsoft.Azure.Functions.Worker.Extensions.Http.AspNetCore.FunctionsHttpProxyingMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in /_/extensions/Worker.Extensions.Http.AspNetCore/src/FunctionsMiddleware/FunctionsHttpProxyingMiddleware.cs:line 38\r\n   at Microsoft.Azure.Functions.Worker.Extensions.DurableTask.DurableTaskFunctionsMiddleware.Invoke(FunctionContext functionContext, FunctionExecutionDelegate next) in /_/src/Worker.Extensions.DurableTask/DurableTaskFunctionsMiddleware.cs:line 31\r\n   at Microsoft.Azure.Functions.Worker.FunctionsApplication.InvokeFunctionAsync(FunctionContext context) in /_/src/DotNetWorker.Core/FunctionsApplication.cs:line 83\r\n   at Microsoft.Azure.Functions.Worker.Handlers.InvocationHandler.InvokeAsync(InvocationRequest request) in /_/src/DotNetWorker.Grpc/Handlers/InvocationHandler.cs:line 91
```

**Comment by linglingye001** [comment]
Hi @kemmis, thanks for providing the package and exception details. The [Durable fix](https://github.com/Azure/azure-functions-durable-extension/pull/3229) first shipped in version `1.11.0`, we tested  `UseAzureAppConfiguration()` with `Microsoft.Azure.Functions.Worker.Extensions.DurableTask` v1.10.0, v1.11.0 and v1.18.0.

**v1.10.0**: repro the original middleware compatibility issue
`System.InvalidOperationException: An invalid asynchronous invocation was detected.
This can be caused by awaiting non-durable tasks in an orchestrator function's
implementation or by middleware that invokes asynchronous code.`

**v1.11.0 and v1.18.0**: The exception was absent, and the orchestration continued normally. This is consistent with the asynchronous middleware fix introduced in DurableTask 1.11.0.

The exception you are seeing appears to be different which indicates that `DurableFunctionExecutor` did not receive the expected orchestration-history input.
`System.InvalidOperationException:
Orchestration history state was either missing from the input or not a string value.`

We have not reproduced that failure with DurableTask 1.18.0, could you please provide additional details? Thanks
